File: warmup/Python/ex6.py
arr = [0, 0, 1, 0, 1, 0, 0]
def longest_sub_binary_array():
    """
    Given a binary array, find the longest subarray where number of 0 == number of 1
    """
    d = {}
    d[0] = -1
    max_len = 0
    end_ind = -1
    add_sum = 0
    for i, num in enumerate(arr):
        if num == 0: add_sum -= 1
        else: add_sum += num
        if add_sum in d:
            s = d[add_sum]
            if i-s > max_len:
                max_len = i-s
                end_ind = i
        else:
            d[add_sum] = i

    print(arr[end_ind-max_len+1:end_ind+1])

longest_sub_binary_array()

    

File: warmup/Java/ex6.java
import java.util.HashMap;
import java.util.Map;

class Ex6 {

    public static void longestSubBinaryArray (int[] nums) {
        // Given a binary array, find the longest subarray where number of 0 === number of 1
        Map<Integer, Integer> d = new HashMap<>();
        d.put(0, -1);
        int maxLen = 0;
        int endInd = 0;
        int addSum = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {addSum--;}
            else {addSum++;}
            if (d.containsKey(addSum)) {
                if (i - d.get(addSum) > maxLen) {
                    maxLen = i - d.get(addSum);
                    endInd = i;
                }
            }
            d.put(addSum, i);
        }
        
    }
    public static void main (String[] args) {
        int[] num = {0, 0, 1, 0, 1, 0, 0};
    }

}